CN Student Dashbaord Load Tests From Database

chrisn02 opened the pull request:
Have made it so that it displays all the assigned tests for the currently logged in user. For the tests to complete, it shows the subject name along with the test name, number of questions and date it was set. The completed test displays similar information, except it shows the date it was completed, the score that was achieved as a percentage and the number of points earned. If there are no tests available, then a message will be displayed saying that there are none available. 

![image](https://github.com/Jforjo/CSD/assets/59711723/29e13d77-407d-4f7d-8a6a-76ce8b29028b) Tests To Complete section with one current test to do

![image](https://github.com/Jforjo/CSD/assets/59711723/9fd021d5-25af-409a-aea6-079821b65091) Once the test has been completed it will be moved to the completed tests section, and a message has appeared saying there are no tests to complete

chrisn02 committed:
Have started loading the tests for the user to complete

Currently does not work properly, will redo completely when I get the chance.

chrisn02 committed:
Am working on loading subject names into testboxes

Can't properly test if it works because there seems to be an issue connecting to the database (target machine actively refused it)

chrisn02 committed:
Now loads the quiz title into the test box

Still need to make it so it only loads the tests of the user currently logged in, but am unable to do so at the moment so it currently just loads every test in the database

chrisn02 committed:
Now loads test data from studentQuizLink table

The test data now loads from the studentQuizLink table. It currently loads every test in the table rather than ones assigned to a specific student as I'm not able to log in yet, but will correct as soon as possible. Still some more things to do though, such as the score being loaded, and having an actual test completion date, rather than just using the date the test was assigned (only done as a placeholder)

chrisn02 committed:
More test box changes

Have added the score for the completed tests by dividing the number of correct answers the user gets by the number of total questions, then multiplying by 100. Have also made it so if there are currently no tests to complete then a message appears saying that there are no tests to complete

chrisn02 committed:
Clicking on start test should load questions in testing page

When you click on the start test button of a test, it takes the ID of the test you clicked on, and then when it goes to the testing page, it should load every question that is linked to that specific test.

chrisn02 committed:
Now only loads quizzes assigned to currently logged in user

Also updated welcome message for currently logged in user

chrisn02 committed:
Now displays user points earned on completed tests

Also some general PHP fixes

chrisn02 committed:
Now gets student data from procedure called "GetStudentData"

chrisn02 committed:
Now calculates the number of questions by seeing how many questions are linked to a quiz in the database

Also cleaned up the SQL query for getting each test as originally there were two separate queries for getting both the completed tests and tests to complete but have made it into one query that then separates them out afterwards

Jforjo commented on dashboard.php:
```diff
@@ -17,8 +17,52 @@
+   $sql = "SELECT quizzes.quizID, quizzes.title, subjects.name, studentQuizLink.TIMESTAMP, studentQuizLink.correctCount, studentQuizLink.questionCount, studentQuizLink.points, studentQuizLink.completed
+   FROM quizzes 
+   JOIN subjects ON quizzes.subjectID = subjects.subjectID
+   JOIN studentQuizLink ON quizzes.quizID = studentQuizLink.quizID AND studentQuizLink.studentID = :studentID";
```
Don't include SQL in the code, this should be within a prepared function/procedure within PHPMyAdmin

chrisn02 replied:
What procedure would I use? I tried "GetAssignedQuiz" and couldn't get it to work

Jforjo replied:
I've just created a procedure to used called `GetStudentsQuizzes`. It returns:
* 'quizID'
* quiz title AS 'quiz'
* subject name AS 'subject'
* 'questionCount'
* 'correctCount'
* 'completed'
* 'points'
* 'dateCompleted'

If there's anymore you'd like added you can either message me or add it yourself.

chrisn02 replied:
Ok thanks. I'll add that to my code now

chrisn02 replied:
Would it be possible to make it so it also returns the TIMESTAMP field of studentQuizLink please? Only because I use that field as the date the quiz was set, unless you think it's not needed.

Jforjo replied:
Done. It now also returns TIMESTAMP AS 'dateSet'.
I've also ordered it by the TIMESTAMP, so the most recently set tests will be at the top of the list.

chrisn02 replied:
Thanks for your help, It now works

Jforjo commented on dashboard.php:
```diff
@@ -67,51 +108,18 @@
         <h2 class="section-title">Completed Tests</h2>
         <!-- Area of page that contains the test boxes -->
         <div class="test-section">
+        <?php foreach ($completedTests as $test): ?>
```
Add a check similar to the Tests To Complete to show a message saying they have no completed tests to show if the `completedTests` is empty.

Jforjo commented on dashboard.php:
```diff
@@ -17,8 +17,52 @@
+        PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
+        PDO::ATTR_EMULATE_PREPARES   => false,
+    ];
+    $pdo = new PDO($dsn, $user, $pass, $opt);
```
Everything from `$dsn = DB_DSN` to this line should be removed as it does nothing - it's all handled in the `newConn` function.

Jforjo reviewed (changes requested):
You shouldn't be merging the auth and login files.
Also, you should add a logout button that links to '/logout' - the router will handle the rest.
See comments for more details.

chrisn02 committed:
Added message for when there are no completed tests

Also added a logout button onto the nav

chrisn02 committed:
Now calls quiz data from procedure

Calls quiz data from a procedure rather than hard coding an SQL query

gchodancar19 reviewed (approved):
For all the tests completed section it will be good to have a view results button as students will be able to see on which parts they have scored highest and on which parts they have scored lowest.

chrisn02 commented:
> For all the tests completed section it will be good to have a view results button as students will be able to see on which parts they have scored highest and on which parts they have scored lowest.

This will be shown on stats page, but I *could* make it so you can see your correct and incorrect answers for each quiz if you click on it

Jforjo commented:
> For all the tests completed section it will be good to have a view results button as students will be able to see on which parts they have scored highest and on which parts they have scored lowest.

We are not storing every answer of every student for every quiz they do. This is a lot of storage and will only be done if the stakeholders say so.

chrisn02 commented:
> We are not storing every answer of every student for every quiz they do. This is a lot of storage and will only be done if the stakeholders say so.

Yeah good point, I won't add this then.

chrisn02 committed:
Resolved merge conflict

Resolved by renaming dashboard.php to index.php and moving it to pages/student

Also removed temporary auth.php and login.php files

chrisn02 committed:
Resolved merge conflict (didn't work properly the first time)

chrisn02 committed:
Merge branch 'sudo-main' into CN-Dashboard-LoadTestData

chrisn02 commented:
Have made the requested changes. I have replaced the SQL query with a procedure and have added a logout button.

Jforjo commented on pages/student/index.php:
```diff
@@ -4,21 +4,52 @@
+   $stmt->bindValue(":studentID", $studentID, PDO::PARAM_STR);
+   $stmt->execute();
+   $allTests = $stmt->fetchAll();
+   
```
I think it's good practice to set `$conn` to null, or maybe it isn't, but I'd recommend you do so.

chrisn02 replied:
Ok, will take a look at that.

Jforjo commented on pages/student/index.php:
```diff
@@ -4,21 +4,52 @@
-                <li><a href="#">Home</a></li>
-                <li><a href="#">Stats</a></li>
+                <li><a href="index.php">Home</a></li>
+                <li><a href="user-stats.php">Stats</a></li>
```
These 2 lines will need to be '/dashboard' and '/stats' or something eventually so it works with the router, but you can decide on those names later, and it can also be done after all the pages are on sudo-main.

chrisn02 replied:
Alright, will do
